tpu/templates/tpu_film_estimator/trainer.py

```diff
@@ -0,0 +1,266 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import argparse
+import numpy as np
+import tensorflow as tf
+from tensorflow.contrib import summary
+
+n_classes = 10
+
+# Making the filter sizes a global variable so it's eaiser to coordinate
+# between the modulation sub-network and the convolutional classifier
+# sub-network.
+filter_sizes = [32, 64]
+
+# A linear modulation will be applied to every filter/feature map.
+n_film = sum(filter_sizes)
+
+# ## Feature-wise Linear Modulation Layer
+#
+# For details, see [FiLM: Visual Reasoning with a General Conditioning Layer](https://arxiv.org/abs/1709.07871).
+#
+
+class FeaturewiseLinearModulationLayer(tf.layers.Layer):
+    def call(self, input_, gamma, beta):
+        # The user is responsible for having the correct shapes
+        return gamma * input_ + beta
+
+
+# ## The model function
+# 
+# The network consists of two sub-networks:
+#
+# * Label classifier: A feedforward network (here convolutional).  
+#   It is linearly modulated at intermediate outputs.
+#
+# * Modulation: A separate sub-network that learns the modulation parameters.
+#
+
+def model_fn(features, labels, mode, params):
+    x = features['x']
+    modulation_data = features['modulation_data']
+    onehot_labels = tf.one_hot(labels, n_classes)
+
+    global_step = tf.train.get_global_step()
+
+    # In this sample we use dense layers for the modulation sub-network.
+    # Its output has shape (batch_size, 2 * n_film) since each FiLM layer has
+    # two parameters.
+    modulation_hidden = tf.layers.dense(modulation_data, 128, activation=tf.nn.relu)
```

Comment:
tf.layers will be deprecated in TF 2.0, not sure if makes sense to move it to tf.keras ?

Reply:
done, changed the model building code to use tf.keras.layers